File: plugins/corpus-meta-events/android/src/main/java/com/corpusapp/metaevents/CorpusMetaEventsPlugin.java
package com.corpusapp.metaevents;

import android.os.Bundle;

import com.getcapacitor.JSObject;
import com.getcapacitor.Plugin;
import com.getcapacitor.PluginCall;
import com.getcapacitor.PluginMethod;
import com.getcapacitor.annotation.CapacitorPlugin;

import com.facebook.FacebookSdk;
import com.facebook.appevents.AppEventsLogger;

import java.util.Iterator;

// ═══════════════════════════════════════════════════════════════════════════
// APP STORE / PLAY STORE PRIVACY DECLARATIONS — READ BEFORE EDITING
// ═══════════════════════════════════════════════════════════════════════════
// The App Store Connect privacy label (iOS) and Play Data Safety declaration
// (Android) both declare every data type Meta receives as "NOT USED FOR
// TRACKING". That declaration is only true if EVERY ONE of the five rules
// below stays in force. Both stores treat a mismatch between declared and
// actual behavior as a rejection reason (Apple Guideline 5.1.2, Play Data
// Safety policy).
//
// If you change ANY of the five rules below, you MUST update BOTH store
// privacy declarations FIRST. Not concurrently — first. The declaration
// version applies to the build being reviewed.
//
//   1. NEVER trigger an Android tracking-consent prompt (equivalent of iOS
//      ATT) for advertising ID access, and NEVER declare the AD_ID
//      permission in the Corpus AndroidManifest.
//
//   2. NEVER call FacebookSdk.setAdvertiserIDCollectionEnabled(true).
//      This is Meta's own switch that tells their pipeline "this install
//      is opted in to advertising-ID collection." Must stay false.
//
//   3. NEVER expose AppEventsLogger.setUserData(...) or any Advanced
//      Matching entry point from this plugin's JS surface. Advanced Matching
//      passes hashed email/phone to Meta so they can join the app-events
//      stream to a Meta profile — that's tracking by any definition.
//
//   4. NEVER enable auto-log for in-app-purchase / subscription events at
//      the SDK level (there is no direct Java equivalent of iOS's
//      isAutoLogSubscriptionsEnabled today, but Meta occasionally adds
//      similar hooks; check release notes on every SDK version bump).
//      Subscription events must flow through this plugin's logEvent() only.
//
//   5. NEVER add a second SDK anywhere in the app that requests advertising
//      consent (AppsFlyer, Adjust, Branch, etc.) — even one such prompt
//      flips the whole-app privacy declaration, regardless of what this
//      plugin does.
//
// The load() call below sets the collection flag to false BEFORE SDK
// initialization. Ordering matters: if any event is logged before the
// flag is set, Meta's SDK reads the pre-set default (which favors
// collection, not privacy).
//
// Test after any config change: on a fresh install, no advertising
// consent prompt should appear; Meta Events Manager should show events as
// "unidentified" (no linked advertiser id / Meta profile id).
// ═══════════════════════════════════════════════════════════════════════════

@CapacitorPlugin(name = "CorpusMetaEvents")
public class CorpusMetaEventsPlugin extends Plugin {

    private AppEventsLogger logger;

    @Override
    public void load() {
        // Rule 2 enforced HERE, before SDK init.
        FacebookSdk.setAdvertiserIDCollectionEnabled(false);
        // Auto-log app-activation events is OK — those don't carry the
        // advertising ID once the flag above is set. Auto-log purchases /
        // subscriptions is NOT ok (rule 4).
        FacebookSdk.setAutoLogAppEventsEnabled(true);
        // App ID + client token are wired via the host AndroidManifest
        // <meta-data> entries. sdkInitialize picks them up automatically.
        FacebookSdk.sdkInitialize(getContext());
        AppEventsLogger.activateApp(getActivity().getApplication());
        logger = AppEventsLogger.newLogger(getContext());
    }

    @PluginMethod
    public void logEvent(PluginCall call) {
        String name = call.getString("name");
        if (name == null || name.isEmpty()) {
            call.reject("missing event name");
            return;
        }
        JSObject params = call.getObject("params");
        Bundle bundle = new Bundle();
        Double valueToSum = null;
        if (params != null) {
            Iterator<String> keys = params.keys();
            while (keys.hasNext()) {
                String k = keys.next();
                Object v = params.opt(k);
                if ("_valueToSum".equals(k) && v instanceof Number) {
                    valueToSum = ((Number) v).doubleValue();
                } else if (v instanceof String) {
                    bundle.putString(k, (String) v);
                } else if (v instanceof Integer) {
                    bundle.putInt(k, (Integer) v);
                } else if (v instanceof Long) {
                    bundle.putLong(k, (Long) v);
                } else if (v instanceof Double) {
                    bundle.putDouble(k, (Double) v);
                } else if (v instanceof Boolean) {
                    bundle.putBoolean(k, (Boolean) v);
                } else if (v != null) {
                    bundle.putString(k, v.toString());
                }
            }
        }
        if (valueToSum != null) {
            logger.logEvent(name, valueToSum, bundle);
        } else if (!bundle.isEmpty()) {
            logger.logEvent(name, bundle);
        } else {
            logger.logEvent(name);
        }
        JSObject ret = new JSObject();
        ret.put("logged", true);
        call.resolve(ret);
    }
}

File: plugins/corpus-meta-events/ios/Sources/CorpusMetaEventsPlugin/CorpusMetaEventsPlugin.swift
import Capacitor
import FBSDKCoreKit
import Foundation
import UIKit

// ═══════════════════════════════════════════════════════════════════════════
// APP STORE CONNECT PRIVACY LABEL COMMITMENT — READ BEFORE EDITING
// ═══════════════════════════════════════════════════════════════════════════
// The App Store Connect privacy label for Corpus declares every data type
// Meta receives as "NOT USED FOR TRACKING". That declaration is only true if
// EVERY ONE of the five rules below stays in force. Apple treats a mismatch
// between the declared label and actual behavior as a Guideline 5.1.2 (Data
// Use and Sharing) rejection reason.
//
// If you change ANY of the five rules below, you MUST update the App Store
// Connect privacy label declaration FIRST. Not concurrently — first. The
// label version applies to the build being reviewed.
//
//   1. NEVER call ATTrackingManager.requestTrackingAuthorization anywhere
//      in the app. Not from this plugin, not from any other plugin, not
//      from JS. One prompt anywhere in the app flips the label.
//
//   2. NEVER set Settings.shared.isAdvertiserTrackingEnabled = true.
//      This is Meta's own switch that tells their pipeline "this install
//      is opted in to cross-app correlation". Must stay false.
//
//   3. NEVER set Settings.shared.isAdvertiserIDCollectionEnabled = true.
//      Turning this on lets Meta read the IDFA (via ATT-approved installs)
//      OR the IDFV as a persistent identifier — either way, "used for
//      tracking" becomes true.
//
//   4. NEVER call AppEvents.shared.setUserData(...) — that's Advanced
//      Matching, which passes hashed email/phone to Meta so they can join
//      the app-events stream to a Meta profile. The JS surface of this
//      plugin deliberately does not expose that method; adding it back
//      would violate the "not used for tracking" commitment.
//
//   5. NEVER enable Settings.shared.isAutoLogSubscriptionsEnabled OR any
//      other auto-log flag that hooks SKPaymentQueue. Subscription events
//      must flow through this plugin's logEvent() only, so the payload is
//      what our code puts in it — not whatever Meta's SDK decides to
//      autofill (which historically has included receipt data with
//      identifiers on some iOS versions).
//
// The plugin.load() below sets all three Meta-side flags to their
// safe values BEFORE SDK initialization. Ordering matters: if any event
// is logged before these flags are set, Meta's SDK reads the pre-set
// defaults (which favor tracking, not privacy).
//
// Test after any config change: on a fresh install, no ATT prompt should
// appear; Meta Events Manager should show events as "unidentified" (no
// linked advertiser id / Meta profile id).
// ═══════════════════════════════════════════════════════════════════════════

@objc(CorpusMetaEventsPlugin)
public class CorpusMetaEventsPlugin: CAPPlugin, CAPBridgedPlugin {
    public let identifier = "CorpusMetaEventsPlugin"
    public let jsName = "CorpusMetaEvents"
    public let pluginMethods: [CAPPluginMethod] = [
        CAPPluginMethod(name: "logEvent", returnType: CAPPluginReturnPromise)
    ]

    public override func load() {
        // Rules 2 + 3 + 5 enforced HERE, before SDK init.
        Settings.shared.isAdvertiserTrackingEnabled = false
        Settings.shared.isAdvertiserIDCollectionEnabled = false
        // Auto-log app-activation events is OK — those don't carry IDFA once
        // the three flags above are set. Auto-log subscription events is
        // NOT ok (rule 5) — Meta's SKPaymentQueue hook can attach receipt
        // data with identifiers.
        Settings.shared.isAutoLogAppEventsEnabled = true
        // SKAdNetwork reporting: opt in. This is Apple's privacy-preserving
        // attribution channel and is orthogonal to IDFA/tracking. Meta reads
        // aggregated conversion signals from Apple, never per-user data.
        Settings.shared.isSKAdNetworkReportEnabled = true

        // Initialize the SDK after flags are set. The app-level
        // ApplicationDelegate.application(_:didFinishLaunchingWithOptions:)
        // call happens from AppDelegate.swift (Capacitor host) — this
        // load() runs during Capacitor bridge init, which precedes the
        // JS runtime by design, so our flags are always the ones in
        // force when any event fires.
        ApplicationDelegate.shared.initializeSDK()
    }

    @objc func logEvent(_ call: CAPPluginCall) {
        guard let name = call.getString("name"), !name.isEmpty else {
            call.reject("missing event name")
            return
        }
        // params is a JS object; Capacitor bridges it as [String: Any].
        // Meta's AppEvents expects [AppEvents.ParameterName: Any] for the
        // typed API or [String: Any] for the raw logEvent — we use the
        // string variant so custom params flow through cleanly.
        let params = call.getObject("params") ?? [:]
        // _valueToSum is a Meta convention: when present, it's the
        // numeric value the event represents (e.g. subscription price).
        // Extract it if the caller passed it, then pass everything else
        // as parameters.
        var extractedValue: Double? = nil
        var otherParams: [AppEvents.ParameterName: Any] = [:]
        for (k, v) in params {
            if k == "_valueToSum", let n = v as? Double {
                extractedValue = n
            } else if k == "_valueToSum", let n = v as? Int {
                extractedValue = Double(n)
            } else {
                otherParams[AppEvents.ParameterName(k)] = v
            }
        }
        let eventName = AppEvents.Name(name)
        if let value = extractedValue {
            AppEvents.shared.logEvent(eventName, valueToSum: value, parameters: otherParams)
        } else if !otherParams.isEmpty {
            AppEvents.shared.logEvent(eventName, parameters: otherParams)
        } else {
            AppEvents.shared.logEvent(eventName)
        }
        call.resolve(["logged": true])
    }
}
